Add PHPStan analysis Github action to the WooCommerce plugin

louwie17 opened the pull request:
### Changes proposed in this Pull Request:

This is a follow up from this exploration pf59Ax-3os-p2 around adding PHPStan analysis to WooCommerce.
This adds an initial config and baseline with a GH action to run on PRs that changed PHP files.
This is currently locked to PHP 7.4, but it appears that in order to update the baseline correctly, you did also have to run it locally using PHP 7.4.

**Note:**

I initially wanted to add this to the `/tools` directory, but setting it up from there didn't work correctly for me, likely requiring some more tooling with the `neon` config file in order to get the correct context.
I resorted to updating it as part of the `WooCommerce` plugin folder.

<details>

<summary>Details of the baseline file</summary>

# PHPStan Baseline Analysis

## Top 5 Most Common Errors

1. **`argument.type`** - 5,534 errors
   - Type mismatches when passing arguments to functions
   - Most common issue: Passing `mixed` where specific types are expected

2. **`missingType.return`** - 4,328 errors
   - Methods missing return type declarations
   - Many methods throughout the codebase lack explicit return types

3. **`class.notFound`** - 1,871 errors
   - PHPStan cannot find class definitions
   - Often related to namespaced classes or WordPress core classes

4. **`method.nonObject`** - 1,780 errors
   - Calling methods on non-object types (e.g., `mixed`, unions that include `bool`)
   - Common pattern: Calling methods on `WC_Order|bool` or `Note|true`

5. **`offsetAccess.nonOffsetAccessible`** - 1,232 errors
   - Accessing array offsets on `mixed` or non-array types
   - Common pattern: `Cannot access offset 0 on mixed`

---

## 10 Suggestions to Drastically Reduce Baseline Size (Low-Hanging Fruit)

### 1. Add Return Type Declarations (4,328 errors)
- **Action**: Add explicit return types to all methods that are missing them
- **Priority**: Start with high-traffic classes like `WC_Data` and its subclasses
- **Impact**: Could eliminate ~4,328 errors

### 2. Add Type Hints for `mixed` Parameters (5,534 errors)
- **Action**: Replace `mixed` with specific types wherever possible
- **Priority**: Focus on frequently used functions:
  - `wc_clean()` (235 errors) - expects `array|string`
  - `absint()` (215 errors) - expects `array|bool|float|int|resource|string|null`
  - `sanitize_text_field()` (157 errors) - expects `string`
- **Impact**: Could eliminate ~1,000+ errors

### 3. Add PHPStan Stubs for Missing Classes (1,871 errors)
- **Action**: Create stub files for classes that PHPStan cannot find
- **Priority**: Focus on:
  - `Automattic\WooCommerce\Blocks\BlockTypes\WP_Block` (68 errors)
  - `Automattic\WooCommerce\Admin\API\Reports\DateTime` (53 errors)
  - `WP_CLI` (multiple errors)
- **Impact**: Could eliminate ~1,871 errors

### 4. Add Type Guards for Union Types with `bool`/`true` (1,780 errors)
- **Action**: Add null/type checks before calling methods on union types like `WC_Order|bool` or `Note|true`
- **Example**: Check `if ($order instanceof WC_Order)` before calling `$order->get_id()`
- **Impact**: Could eliminate ~500-800 errors

### 5. Add Array Type Hints for Offset Access (1,232 errors)
- **Action**: Type arrays explicitly instead of `mixed` before accessing offsets
- **Priority**: Use `@var array` or `array<string, mixed>` annotations
- **Impact**: Could eliminate ~1,232 errors

### 6. Add Constant Definitions (266+ errors)
- **Action**: Define missing constants or add them to PHPStan stubs
- **Priority**: 
  - `WC_ABSPATH` (266 errors)
  - `WC_ADMIN_IMAGES_FOLDER_URL` (73 errors)
- **Impact**: Could eliminate ~339+ errors

### 7. Fix `call_user_func` Type Issues (104 errors)
- **Action**: Type callbacks explicitly instead of passing `mixed`
- **Priority**: Use `callable` type hints or specific callable signatures
- **Impact**: Could eliminate ~104 errors

### 8. Add Null Checks for Potentially Undefined Variables (101+ errors)
- **Action**: Add checks for variables that might not be defined
- **Priority**: 
  - `$order might not be defined` (101 errors)
  - `$loop might not be defined` (54 errors)
- **Impact**: Could eliminate ~155+ errors

### 9. Fix `@param` Tag Issues (52+ errors)
- **Action**: Replace `@param $this` with descriptive aliases like `@param $instance`
- **Impact**: Could eliminate ~52+ errors

### 10. Add Type Casts with Validation (85+ errors)
- **Action**: Replace unsafe casts with validated conversions
- **Example**: Instead of `(int) $mixed`, use `is_numeric($mixed) ? (int) $mixed : 0`
- **Impact**: Could eliminate ~85+ errors

---

## Estimated Impact

Addressing these 10 suggestions could potentially reduce the baseline by approximately **10,000-12,000 entries** (about 9-11% of the current 108,033 lines), with the biggest wins coming from:

- Adding return types (4,328 errors)
- Fixing `mixed` type issues (5,534 errors)
- Adding class stubs (1,871 errors)

These are all relatively straightforward fixes that can be tackled incrementally.
</details>


### Screenshots or screen recordings:

<img width="760" height="80" alt="Screenshot 2025-11-21 at 14 17 56" src="https://github.com/user-attachments/assets/04b30a38-3482-4a36-be8f-6c16cba42566" />

<img width="1205" height="784" alt="Screenshot 2025-11-21 at 14 30 28" src="https://github.com/user-attachments/assets/114a8bf5-fc6b-4f1d-8f4d-4b7c5e02a8ca" />

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:
**Requirements: **
- Install PHP 7.4
- Run `export PATH="PATH_TO_PHP_7.4:$PATH"`
- `php -v` should output 7.4 version.

1. Go to **plugins/woocommerce** and run `composer i`
2. Run `composer phpstan` and see if it runs correctly without errors ( if nothing drastic changed with after a rebase )
3. Run `pnpm --filter=@woocommerce/plugin-woocommerce build:zip`
4. Move the `woocommerce.zip` to a new folder and unzip it
5. Make sure that the phpstan related neon files, and `php-stub` directory are **not** included.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

louwie17 committed:
Add phpstan analysis

louwie17 committed:
Update config and baseline

louwie17 committed:
Add Github action for PHPStan analysis

louwie17 committed:
Add phpstan files to the distignore

louwie17 committed:
Fix missing feature config file when using phpstan

louwie17 committed:
Add optional to feature config path

louwie17 committed:
Update memory limit and processes

louwie17 committed:
Skip php stubs with linting and update README

louwie17 committed:
Add changelog

louwie17 commented on plugins/woocommerce/php-stubs/wc-admin-feature-config.php:
```diff
@@ -0,0 +1,14 @@
+<?php
```
Created a `stub` for this as it is `.gitignored` and only generated on `build`

louwie17 commented on plugins/woocommerce/phpstan-baseline.neon:
Generated baseline file based off of level 9 and the current repo structure. I left a summary from Cursor in the PR description "Details of the baseline file"

louwie17 commented on plugins/woocommerce/phpstan.neon:
```diff
@@ -0,0 +1,27 @@
+	excludePaths:
+		# Matches the prior test implementation; GeoIP relies on data files.
+		- includes/class-wc-geo-ip.php
+		- includes/react-admin/feature-config.php (?)
```
Exclude this as it is `gitignored`, also added the `(?)` as it can exist locally.

Nikschavan commented on plugins/woocommerce/phpstan.neon:
```diff
@@ -0,0 +1,27 @@
+parameters:
+	tmpDir: tmp
+	level: 9
+	phpVersion: 80400
```
Backwards Compatibility Review

**Status:** ✅ No BC issue - This was corrected

**Analysis:** The PHPStan configuration correctly sets `phpVersion: 70400` (PHP 7.4), which aligns with WooCommerce's minimum PHP requirement and the GitHub workflow configuration. The initial concern was based on an incorrect reading of the value.

**Verification:**
- `phpstan.neon:7` sets `phpVersion: 70400` ✓
- `.github/workflows/phpstan.yml:42` uses `php-version: '7.4'` ✓
- Both configs are consistent and match WooCommerce's support policy

<!-- wc-bc:path=plugins/woocommerce/phpstan.neon;ol=7;topic=php-version-constraint -->

louwie17 replied:
Good suggestion, I updated it :)

Nikschavan commented on .github/workflows/phpstan.yml:
```diff
@@ -0,0 +1,65 @@
+              with:
+                  install: '@woocommerce/plugin-woocommerce...'
+                  pull-package-deps: '@woocommerce/plugin-woocommerce'
+                  php-version: '8.4'
```
Same as [other comment](https://github.com/woocommerce/woocommerce/pull/62064/files#r2555801077), should this be run on a lower PHP version?

louwie17 replied:
Yeah I updated it 👍

Nikschavan commented on plugins/woocommerce/composer.json:
```diff
@@ -153,6 +157,12 @@
+			"vendor/bin/phpstan analyse --configuration=phpstan.neon --memory-limit=4G"
+		],
+		"phpstan:baseline": [
+			"vendor/bin/phpstan analyse --configuration=phpstan.neon --memory-limit=4G --generate-baseline=phpstan-baseline.neon"
```
While this does make updating, one thing of caution is that engineers running this command may accidentally ignore new errors - I am not saying we should not have this command, but this is something to document and inform when this gets merged and published?

Nikschavan reviewed (approved):
This tested well. I have added a question regarding the PHP Version used in the PHPStan configuration.

The PHPStan command runs as expected.

<img width="1794" height="322" alt="Antigravity -2025-11-24 at 16 47 27@2x" src="https://github.com/user-attachments/assets/730c8de9-1baa-48fe-81ba-22efb4002492" />

chihsuan commented on plugins/woocommerce/composer.json:
```diff
@@ -153,6 +157,12 @@
 		"build-lib": [
 			"sh ./bin/build-lib.sh"
+		],
+		"phpstan": [
```
Since we typically use `pnpm` to run our commands, should we consider adding these commands to `package.json` as well?

louwie17 replied:
Updated!

chihsuan commented on .github/workflows/phpstan.yml:
```diff
@@ -0,0 +1,65 @@
+
+            - name: 'Run PHPStan'
+              working-directory: plugins/woocommerce
+              run: composer phpstan
```
Just wanted to share that in the Jetpack monorepo, CI runs PHPStan analysis using the `--update-baseline` flag and reports whether the baseline has improved or regressed. From my experience, it's helpful to provide contributors with clear next steps when the PHPStan analysis fails in CI. If it's easy to implement, maybe we can do the same here.

https://github.com/Automattic/jetpack/blob/02fb5055e718f06ffa8a3fdb15db3e7fd020ee46/.github/workflows/tests.yml#L491-L541

louwie17 replied:
Thanks that is a good suggestion, I created a follow up PR to add this: https://github.com/woocommerce/woocommerce/pull/62132

triple0t commented on packages/php/email-editor/tasks/phpstan/phpstan.neon:
```diff
@@ -1,6 +1,6 @@
 parameters:
 	level: 9
-	phpVersion: 80400
+	phpVersion: 70400
```
Please don't change this.

The email editor [runs phpstan](https://github.com/woocommerce/woocommerce/blob/8ca3fd91b598e3e22e5071e78923453b3dcbdff8/packages/php/email-editor/package.json#L60-L83) analysis for both PHP 7 and PHP 8

php7 config: https://github.com/woocommerce/woocommerce/blob/8ca3fd91b598e3e22e5071e78923453b3dcbdff8/packages/php/email-editor/tasks/phpstan/phpstan-7.neon

php8 config: https://github.com/woocommerce/woocommerce/blob/8ca3fd91b598e3e22e5071e78923453b3dcbdff8/packages/php/email-editor/tasks/phpstan/phpstan.neon

louwie17 replied:
Thanks, this was changed by accident, my bad!

louwie17 committed:
Update to use PHP 7.4

louwie17 committed:
Use pnpm to run phpstan instead

louwie17 committed:
Update baseline

louwie17 committed:
Update baseline with PHP 7.4 run

louwie17 committed:
Remove extension installer and load extension directly

albarin reviewed (approved):
✅ Confirmed `composer phpstan` runs without errors.
✅ Confirmed `phpstan` related files and the `php-stub` directory are not included in the zip.

Nikschavan committed:
Update composer.lock to sync with composer.json after rebase

- Update php-stubs/wordpress-stubs constraint to ^6.8 and lock to v6.8.3
- Update woocommerce/monorepo-plugin to current branch
- Remove phpstan/extension-installer (per previous commit)

Nikschavan committed:
Use newer PHP version when running phpstan

PHPStan can be configured to run on older PHP versions, but the installed PHP binary can be of a newer version

PHPStan internally assesses the code with older syntax.

This allows us to update to a newer version of - php-stubs/wordpress-stubs

Nikschavan commented on plugins/woocommerce/composer.json:
```diff
@@ -153,6 +158,12 @@
+			"vendor/bin/phpstan analyse --configuration=phpstan.neon --memory-limit=4G"
+		],
+		"phpstan:baseline": [
+			"vendor/bin/phpstan analyse --configuration=phpstan.neon --memory-limit=4G --generate-baseline=phpstan-baseline.neon"
```
This is being addressed in the follow-up PR https://github.com/woocommerce/woocommerce/pull/62132, The Job will fail if new errors are added to the baseline file

chihsuan reviewed (approved):
LGTM!

chihsuan committed:
Merge branch 'trunk' into add/phpstan

chihsuan committed:
Merge branch 'trunk' into add/phpstan

chihsuan committed:
Update phpstan-baseline.neon
